lib/galaxy/tools/actions/__init__.py

```diff
@@ -881,13 +881,31 @@ def _wrapped_params(self, trans, tool, incoming, input_datasets=None):
         wrapped_params = WrappedParameters(trans, tool, incoming, input_datasets=input_datasets)
         return wrapped_params
 
-    def _get_on_text(self, inp_data):
+    def _get_on_text(self, inp_data, inp_dataset_collections):
         input_names = []
+        collection_names = []
+        collection_hda_ids = set()
+        # output collection id and store included hda ids (to avoid extra inclusion in the list of datasets)
+        # two for loops because:
+        # - inp_dataset_collections stores the collections per input parameter in a dict
+        # - the values contain a list of pairs of DatasetCollectionsAssociations and a bool
+        for collections in inp_dataset_collections.values():
+            for dataset_collection, _ in collections:
+                if getattr(dataset_collection, "hid", None):
+                    collection_names.append(f"{dataset_collection.hid}")
+                if (
+                    getattr(dataset_collection, "collection", None) is None
+                    or getattr(dataset_collection.collection, "elements", None) is None
+                ):
+                    continue
+                for element in dataset_collection.collection.elements:
+                    collection_hda_ids.add(element.hda_id)
```

Comment:
This is a pretty expensive way of getting all hda ids. If you do this only to skip listing the collection and inp_data ... those are keyed on the input. Just use that ?